Resolve logical viewport units properly in font-size

nt1m opened the pull request:
#### 5cd7506926c2b38de056ec7dec5c62d1a049c052
<pre>
Resolve logical viewport units properly in font-size
<a href="https://bugs.webkit.org/show_bug.cgi?id=238982">https://bugs.webkit.org/show_bug.cgi?id=238982</a>

Reviewed by Antti Koivisto.

Logical viewport units depend on the current element&apos;s writing mode, and they can be used in font-size, so that means we need to:

- Make sure writing-mode is always applied first and introduce a new top-priority property type
- Stop passing in unconditionally the parentStyle into CSSToLengthConversionData, since viewport units use the current element&apos;s writing mode, not the parent element&apos;s

* LayoutTests/TestExpectations:
* Source/WebCore/css/CSSPrimitiveValue.cpp:
(WebCore::CSSPrimitiveValue::computeNonCalcLengthDouble):
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/CSSToLengthConversionData.h:
(WebCore::CSSToLengthConversionData::fontCascadeForFontUnits const):
(WebCore::CSSToLengthConversionData::computedLineHeightForFontUnits const):
(WebCore::CSSToLengthConversionData::copyForFontSize const):
(WebCore::CSSToLengthConversionData::copyForFontSizeWithParentStyle const): Deleted.
* Source/WebCore/css/makeprop.pl:
(addProperty):
(sortByDescendingPriorityAndName):
* Source/WebCore/style/StyleBuilder.cpp:
(WebCore::Style::Builder::applyAllProperties):
(WebCore::Style::Builder::applyTopPriorityProperties):
(WebCore::Style::Builder::applyHighPriorityProperties):
* Source/WebCore/style/StyleBuilderCustom.h:
(WebCore::Style::BuilderCustom::applyValueFontSize):
* Source/WebCore/style/StyleResolver.cpp:
(WebCore::Style::Resolver::applyMatchedProperties):
* Tools/Scripts/webkitpy/style/checkers/jsonchecker.py:

Canonical link: <a href="https://commits.webkit.org/253087@main">https://commits.webkit.org/253087@main</a>
</pre>

anttijk commented on Source/WebCore/css/CSSToLengthConversionData.h:
```diff
@@ -65,15 +65,16 @@ class CSSToLengthConversionData {
     const RenderView* renderView() const { return m_renderView; }
     const Element* elementForContainerUnitResolution() const { return m_elementForContainerUnitResolution.get(); }
 
+    const RenderStyle* styleForFontUnits() const { return computingFontSize() ? parentStyle() : style(); }
```
I think this could be narrowed to fontCascadeForFontUnits/computedLineHeightForFontUnits. Nothing else should be coming from the parent style.

anttijk commented on Source/WebCore/style/StyleBuilder.cpp:
```diff
@@ -92,6 +92,9 @@ void Builder::applyAllProperties()
+    // writing-mode defines vi/vb units which may be used in other high priority properties
+    applyProperties(CSSPropertyWritingMode, CSSPropertyWritingMode);
+
     applyProperties(CSSPropertyWebkitRubyPosition, CSSPropertyWebkitRubyPosition);
```
We apply these twice. Can we avoid that?

nt1m replied:
Fixed

dcrousso commented on Source/WebCore/css/CSSPrimitiveValue.cpp:
```diff
@@ -945,12 +945,12 @@ double CSSPrimitiveValue::computeNonCalcLengthDouble(const CSSToLengthConversion
         if (conversionData.computingLineHeight() || conversionData.computingFontSize()) {
             // Try to get the parent's computed line-height, or fall back to the initial line-height of this element's font spacing.
-            value *= conversionData.parentStyle() ? conversionData.parentStyle()->computedLineHeight() : conversionData.style()->metricsOfPrimaryFont().lineSpacing();
+            value *= conversionData.parentStyle() ? conversionData.parentStyle()->computedLineHeight() : conversionData.styleForFontUnits()->metricsOfPrimaryFont().lineSpacing();
```
Does this actually need to change? We've already checked `conversionData.parentStyle()`, so it seems like the `parentStyle()` inside `styleForRontUnits` would be `nullptr`.  Or maybe this is a different value?

dcrousso commented on Source/WebCore/css/CSSPrimitiveValue.cpp:
```diff
@@ -838,16 +838,16 @@ double CSSPrimitiveValue::computeNonCalcLengthDouble(const CSSToLengthConversion
     case CSSUnitType::CSS_QUIRKY_EMS:
-        ASSERT(conversionData.style());
-        value = computeUnzoomedNonCalcLengthDouble(primitiveType, value, conversionData.propertyToCompute(), nullptr, &conversionData.style()->fontDescription());
+        ASSERT(conversionData.styleForFontUnits());
```
Can we actually guarantee that `styleForFontUnits()` is valid (i.e. that we have a `parentStyle()`) no matter how we get here?

anttijk commented on Source/WebCore/style/StyleBuilder.cpp:
```diff
@@ -92,6 +92,9 @@ void Builder::applyAllProperties()
+    // writing-mode defines vi/vb units which may be used in other high priority properties
+    applyProperties(CSSPropertyWritingMode, CSSPropertyWritingMode);
+
     applyProperties(CSSPropertyWebkitRubyPosition, CSSPropertyWebkitRubyPosition);
```
It would be nice if these were applied based on the "top-priority" attribute in json ("applyTopPriorityProperties()")

nt1m committed:
Resolve logical viewport units properly in font-size
https://bugs.webkit.org/show_bug.cgi?id=238982

Reviewed by Antti Koivisto.

Logical viewport units depend on the current element's writing mode, and they can be used in font-size, so that means we need to:

- Make sure writing-mode is always applied first and introduce a new top-priority property type
- Stop passing in unconditionally the parentStyle into CSSToLengthConversionData, since viewport units use the current element's writing mode, not the parent element's

* LayoutTests/TestExpectations:
* Source/WebCore/css/CSSPrimitiveValue.cpp:
(WebCore::CSSPrimitiveValue::computeNonCalcLengthDouble):
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/CSSToLengthConversionData.h:
(WebCore::CSSToLengthConversionData::fontCascadeForFontUnits const):
(WebCore::CSSToLengthConversionData::computedLineHeightForFontUnits const):
(WebCore::CSSToLengthConversionData::copyForFontSize const):
(WebCore::CSSToLengthConversionData::copyForFontSizeWithParentStyle const): Deleted.
* Source/WebCore/css/makeprop.pl:
(addProperty):
(sortByDescendingPriorityAndName):
* Source/WebCore/style/StyleBuilder.cpp:
(WebCore::Style::Builder::applyAllProperties):
(WebCore::Style::Builder::applyTopPriorityProperties):
(WebCore::Style::Builder::applyHighPriorityProperties):
* Source/WebCore/style/StyleBuilderCustom.h:
(WebCore::Style::BuilderCustom::applyValueFontSize):
* Source/WebCore/style/StyleResolver.cpp:
(WebCore::Style::Resolver::applyMatchedProperties):
* Tools/Scripts/webkitpy/style/checkers/jsonchecker.py:

Canonical link: https://commits.webkit.org/253087@main
